### PR TITLE
introduce RegularFileProperty for output in WriteProperties Task

### PR DESCRIPTION
**Pull request by reinsch82**
- resolves: #13363

Signed-off-by: Reinhold Degenfellner <rdegenfellner@gradle.com>

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java [hunk 1]
```diff
@@ -191,8 +195,12 @@ public void setEncoding(String encoding) {
      * Returns the output file to write the properties to.
      */
     @OutputFile
+    @Optional
     public File getOutputFile() {
-        return getServices().get(FileOperations.class).file(outputFile);
+        if(outputFile != null) {
+            return getServices().get(FileOperations.class).file(outputFile);
+        }
+        return null;
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java:203]: :x: This introduce a disconnected new property. You should instead have the old one delegates to the new one. e.g.
```suggestion
        return destinationFile.getAsFile().getOrNull();
```

and have the setter for `outputFile` actually call `destinationFile.set(...)`.

**Review by eskatos — CHANGES_REQUESTED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by eskatos — APPROVED** [review]
LGTM! :+1:

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle clean

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/WritePropertiesIntegrationTest.groovy [hunk 2]
```diff
@@ -45,32 +46,48 @@ class WritePropertiesIntegrationTest extends AbstractIntegrationSpec {
...
     }
 
-    def "simple properties are written sorted alphabetically"() {
+    private runProps() {
+        result = expectOutputDeprecation(executer.withTasks("props")).run()
+    }
+
+    private expectOutputDeprecation(GradleExecuter runWithTasks) {
+        runWithTasks.expectDocumentedDeprecationWarning("he WriteProperties_Decorated.outputFile property has been deprecated. This is scheduled to be removed in Gradle 9.0. " +
+            "Please use the destinationFile property instead. See https://docs.gradle.org/current/dsl/org.gradle.api.tasks.WriteProperties_Decorated.html#org.gradle.api.tasks.WriteProperties_Decorated:outputFile for more details.")
```
Comment (eskatos) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/WritePropertiesIntegrationTest.groovy:62]: :x: The documentation link is invalid. It should point to `https://docs.gradle.org/current/dsl/org.gradle.api.tasks.WriteProperties.html` without the `_Decorated`

File: subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java [hunk 3]
```diff
@@ -191,30 +196,58 @@ public void setEncoding(String encoding) {
...
      */
     @OutputFile
+    @Optional
+    @Deprecated
     public File getOutputFile() {
-        return getServices().get(FileOperations.class).file(outputFile);
+        DeprecationLogger.deprecateProperty(getClass(), "outputFile").replaceWith("destinationFile")
+            .willBeRemovedInGradle9()
+            .withDslReference()
+            .nagUser();
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java:205]: :x: fix the link, see comment above, and extract deprecation logging in a common function

File: subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java [hunk 4]
```diff
@@ -191,30 +196,58 @@ public void setEncoding(String encoding) {
...
      */
+    @Deprecated
     public void setOutputFile(Object outputFile) {
-        this.outputFile = outputFile;
+        DeprecationLogger.deprecateProperty(getClass(), "outputFile").replaceWith("destinationFile")
+            .willBeRemovedInGradle9()
+            .withDslReference()
+            .nagUser();
+
+        this.getDestinationFile().set(getServices().get(FileOperations.class).file(outputFile));
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java:233]: :lipstick: unnecessary `this`
```suggestion
        getDestinationFile().set(getServices().get(FileOperations.class).file(outputFile));
```

File: subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java [hunk 5]
```diff
@@ -224,6 +257,15 @@ public void writeProperties() throws IOException {
...
     }
 
+    private File getTargetFile() {
+        RegularFileProperty df = getDestinationFile();
+        Provider<File> fileProvider = df.getAsFile();
+        if (fileProvider.isPresent()) {
+            return fileProvider.get();
+        }
+        return getOutputFile();
+    }
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java:267]: :x: This is unnecessary and will trigger the deprecation warning even if one doesn't use the deprecated property.
It should simply be `getDestinationFile().getAsFile().get();`

File: subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java [hunk 6]
```diff
@@ -191,30 +196,58 @@ public void setEncoding(String encoding) {
      * Returns the output file to write the properties to.
      */
     @OutputFile
+    @Optional
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/api/tasks/WriteProperties.java:199]: :x: The output is already tracked by `destinationFile` this should be made internal
```suggestion
    @Internal
```

**Review by eskatos — CHANGES_REQUESTED** [review]
Morning brain works better for reviews. I requested more changes to make this correct.

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by eskatos — CHANGES_REQUESTED** [review]
Code looks good :+1: 

The DSL reference should be edited to add the new property.
This can be done in `subprojects/docs/src/docs/dsl/org.gradle.api.tasks.WriteProperties.xml`.

**Comment by reinsch82** [comment]
@bot-gradle this BD

**Comment by reinsch82** [comment]
@bot-gradle test BD

**Review by eskatos — APPROVED** [review]
:+1:

**Comment by reinsch82** [comment]
@bot-gradle test and merge

**Comment by reinsch82** [comment]
@bot-gradle test and merge

**Pull request merged**